paf-mvp-cmp/src/model.ts

```diff
@@ -0,0 +1,433 @@
+   */
+  public get nonePersisted(): boolean {
+    return this.pref.persisted == null && this.rid.persisted == false;
+  }
+
+  // Fields that are used internally to relate values to one another.
+  readonly allFields: IField[];
+  readonly personalizedFields: FieldSingle[];
+  readonly standardFields: FieldSingle[];
+  readonly customFields: FieldSingle[];
+
+  constructor() {
+    // All the fields. Used for the reset and bind methods.
+    this.allFields = [
+      this.onlyThisSite,
+      this[1],
+      this[2],
+      this[3],
+      this[4],
+      this[5],
+      this[6],
+      this[7],
+      this[8],
+      this[9],
+      this[10],
+      this[11],
+      this[12],
+      this.all,
+      this.canSave,
+      this.rid,
+      this.pref,
+    ];
+
+    // All the custom boolean fields that appear in the persisted data.
+    this.customFields = [
+      this[1],
+      this[2],
+      this[3],
+      this[4],
+      this[5],
+      this[6],
+      this[7],
+      this[8],
+      this[9],
+      this[10],
+      this[11],
+      this[12],
+    ];
+
+    // The custom fields that are set to true when standard marketing is enabled.
+    this.standardFields = [this[1], this[2], this[3], this[4], this[5], this[6], this[11], this[12]];
+
+    // The custom fields that are set to true when personalized marketing is enabled.
+    this.personalizedFields = this.customFields;
+  }
+
+  /**
+   * Calls the bind method on all the fields in the model to connect them to the currently displayed UI.
+   */
+  public bind() {
+    this.allFields.forEach((f) => f.bind());
+  }
+
+  /**
+   * Resets all the fields in the model. Sets the setting values flag to ensure the update others method is not called
+   * because we don't want to change the value of dependent fields during a reset.
+   */
+  public reset() {
+    this.settingValues = true;
+    this.allFields.forEach((f) => f.reset());
+    this.settingValues = false;
+  }
+
+  /**
+   * Resets the model and then sets the values from the ids and preferences provided.
+   * @param data
+   */
+  public setFromIdsAndPreferences(data: IdsAndOptionalPreferences) {
+    this.reset();
+    if (data != undefined) {
+      if (data.identifiers != undefined && data.identifiers.length > 0) {
+        Validate.Identifiers(data.identifiers);
+        this.rid.setPersisted(data.identifiers);
+      }
+      if (data.preferences != undefined) {
+        this.pref.setPersisted(data.preferences);
+      }
+    }
+  }
+}
+
+/**
+ * Static class used to determine if the save button should be enabled.
+ */
+class ButtonState {
+  /**
+   * True if the model is in a state where the data can be saved, otherwise false.
+   * @param model
+   * @returns
+   */
+  static isEnabled(model: Model) {
+    return model.onlyThisSite.value === true || model.pref.hasChanged === true || model.rid.persisted === false;
+  }
+}
+
+/**
+ * Static class used to validate complex data fields.
+ */
+class Validate {
```

Comment:
This should be moved to core as it can be used not only by the CMP.

Reply:
Done.

Reply:
I don't understand, is it in another PR?